### ISSUE TITLE
Unaligned access in TopicInfo serialize function on 32-bit platforms

### ISSUE THREAD

**Issue by danepowell** [ros-drivers/rosserial#10]
I am trying to implement rosserial on an STM32F0 (ARM-based) board. When I try to connect to the board using serial_node.py, the code running on my board segfaults in rosserial_msgs/TopicInfo.h (which I generated using the latest version of make_library.py) at the memcpy functions. If I change my topic name to be 8 characters, I don't get the error. It seems that rosserial doesn't check that writes are block-aligned on 32-bit platforms.

**Comment by danepowell** [comment]
Note that the serialization function in string.h also needs work.

**Comment by danepowell** [comment]
Note that I've worked around this temporarily by [using protocol buffers](https://github.com/danep2/rosserial/tree/protobuf). The maintainer of rosserial has stated that he does not want to incorporate protocol buffers though, so a proper fix is still needed.

**Comment by PaulBouchier** [comment]
I just realized I've seen this problem - thanks for reporting it Dane - I didn't know what caused it. The EmbeddedLinux version of rosserial was throwing alignment error messages to the console. It has to be this problem. It prevents use of rosserial on arm platforms, which is a growing problem - should be a problem on Arduino Due too. Needs to be fixed.

**Comment by chuck-h** [comment]
I am seeing this bug as well on the Cypress PSoC4 micro (Arm Cortex-M0). 

BTW rosserial _does_ run on Arduino Due (e.g. https://github.com/ros-drivers/rosserial/pull/58), probably because its Cortex-M3 processor can do unaligned loads & stores (http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0552a/BABFAIGG.html)

**Comment by mikeferguson** [comment]
M4 devices will also have trouble if gcc generates any instructions using the floating point unit.

**Comment by lorenzoriano** [comment]
I'm having the same issue on an ARMv9. Can anybody point me to some resources where I can learn about this problem? I don't even know what to search for. What are alignment errors and how does one fix them?

Thanks!

**Comment by chuck-h** [comment]
@Lorenzo

Basically an alignment error comes because most processors want big
datatypes (e.g. 4-byte, 32-bit integers) to be stored in memory at
addresses that are exact multiples of their size (e.g. addresses 1000,
1004, 1008 would be ok for 32-bit integers, 1001, 1002, 1003 would not).

A brief note on patching here:
http://wiki.jigsawrenaissance.org/index.php?title=Rosserial_On_Cypress_PSoC4#Partial_success.2C_2-Sep-2013 (look after the paragraph starting "2. The ARM-M0 processor does not
support unaligned memory access...")

**Comment by lorenzoriano** [comment]
@chuck-h thanks for the info!

A workaround is to do:
echo 2 > /proc/cpu/alignment

however, as mentioned in https://www.kernel.org/doc/Documentation/arm/mem_alignment, this is very slow and not a proper solution.

If you compile the code with -Wcast-align you'll see all the problematic lines in the code. Basically it's in TopicInfo.h, RequestParam.h and Log.h and the offending instructions are things like:

```
uint32_t * length_topic_name = (uint32_t *)(outbuffer + offset);
```

where (outbuffer + offset) is a char*. This clearly shows the misalignment.

Now these files are generated from python code and I really don't have enough grasping of it to change it. Does anybody have an idea about how to fix it that doesn't involve protobufs? It's a pity that a module targeted to embedded systems has problems with ARM processors.

**Comment by chuck-h** [comment]
Here's a patch you can try, it changes the serialization/deserialization code generated for String types:
https://github.com/chuck-h/rosserial/commit/f0bf4a758bbac81e6a1f96a601297b25b22614b5

**Comment by lorenzoriano** [comment]
Hi @chuck-h , I've seen your contribution too! I think we're both doing the same in a slight different way. We will see if the maintainer accepts any of two solutions.

**Comment by PaulBouchier** [comment]
merged Lorenzoriano's pull request

**Comment by lorenzoriano** [comment]
Hi @PaulBouchier ,

the pull request doesn't seem to have been merged.

**Comment by PaulBouchier** [comment]
Hi Lorenzo - I've merged it with rosserial_server on a local branch. It
corrected an alignment error I was having on my VexPro (ARM/Linux). (I'm
testing all this stuff). I've got 2 more pulls (plus the fork one if you
can find a fix) to merge in then I can push the whole lot out. Are you in a
big rush? If so, let me know & I'll do an intermediate push.

Paul

**Comment by lorenzoriano** [comment]
No rush, I though there was a problem with GitHub! I'm working on some other bugfixes (my fork is getting all sort of branches), expect more pulls!

**Comment by PaulBouchier** [comment]
Fix released in 0.5.3. New debians should be out in a few days.
